File: app/src/main/java/com/mai/nix/maiapp/navigation_fragments/SettingsFragment.java
package com.mai.nix.maiapp.navigation_fragments;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.preference.ListPreference;
import android.preference.Preference;
import android.preference.PreferenceFragment;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;
import com.mai.nix.maiapp.ChooseGroupActivity;
import com.mai.nix.maiapp.R;
import com.mai.nix.maiapp.UserSettings;
import com.mai.nix.maiapp.WebViewActivity;

/**
 * Created by Nix on 03.08.2017.
 */

public class SettingsFragment extends PreferenceFragment implements android.preference.Preference.OnPreferenceClickListener {
    private Preference mGroupPreference;
    private Preference mClearSubjectsCache;
    private Preference mClearExamsCache;
    private ListPreference mFregSubjects;
    private ListPreference mFregExams;
    private ListPreference mLinks;
    private Preference mAbout;
    private Preference mMAI;
    private Preference mDev;
    private static final int REQUEST_CODE_GROUP = 0;
    private String mChoosenGroup;
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        addPreferencesFromResource(R.xml.pref_test);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        UserSettings.initialize(getActivity());
        mGroupPreference = getPreferenceManager().findPreference("pref_group");
        mClearSubjectsCache = getPreferenceScreen().findPreference("clear_cache_subj");
        mClearExamsCache = getPreferenceScreen().findPreference("clear_cache_ex");
        mFregSubjects = (ListPreference)getPreferenceScreen().findPreference("freg");
        mFregExams = (ListPreference)getPreferenceScreen().findPreference("freg_ex");
        mLinks = (ListPreference) getPreferenceScreen().findPreference("links");

        mLinks.setValue(UserSettings.getLinksPreference(getActivity()));
        mGroupPreference.setSummary(UserSettings.getGroup(getActivity()));
        mFregSubjects.setValue(UserSettings.getSubjectsUpdateFrequency(getActivity()));
        mFregExams.setValue(UserSettings.getExamsUpdateFrequency(getActivity()));

        mAbout = getPreferenceScreen().findPreference("about");
        mMAI = getPreferenceScreen().findPreference("go_mai");
        mDev = getPreferenceScreen().findPreference("go_dev");
        mGroupPreference.setOnPreferenceClickListener(this);
        mClearSubjectsCache.setOnPreferenceClickListener(this);
        mClearExamsCache.setOnPreferenceClickListener(this);
        mAbout.setOnPreferenceClickListener(this);
        mMAI.setOnPreferenceClickListener(this);
        mDev.setOnPreferenceClickListener(this);
        mFregSubjects.setOnPreferenceClickListener(this);
        mFregExams.setOnPreferenceClickListener(this);
        mLinks.setOnPreferenceClickListener(this);
        return super.onCreateView(inflater, container, savedInstanceState);
    }

    @Override
    public boolean onPreferenceClick(android.preference.Preference preference) {
        switch (preference.getKey()){
            case "pref_group":
                Intent i = ChooseGroupActivity.newIntent(getActivity(), false);
                startActivityForResult(i, REQUEST_CODE_GROUP);
                break;
            case "clear_cache_subj":
                Toast.makeText(getActivity(), R.string.clear_cache_subj_toast, Toast.LENGTH_SHORT).show();
                break;
            case "clear_cache_ex":
                Toast.makeText(getActivity(), R.string.clear_cache_exams_toast, Toast.LENGTH_SHORT).show();
                break;
            case "about":
                Toast.makeText(getActivity(), R.string.author, Toast.LENGTH_SHORT).show();
                break;
            case "go_mai":
                Uri uri = Uri.parse("http://mai.ru/");
                if(UserSettings.getLinksPreference(getActivity()).equals(UserSettings.ONLY_BROWSER)){
                    Intent intent = new Intent(Intent.ACTION_VIEW, uri);
                    startActivity(intent);
                }else{
                    Intent intent = WebViewActivity.newInstance(getActivity(), uri);
                    startActivity(intent);
                }
                break;
            case "go_dev":
                Toast.makeText(getActivity(), R.string.not_now, Toast.LENGTH_SHORT).show();
                break;
        }
        return true;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode != Activity.RESULT_OK) {
            return;
        }
        if (requestCode == REQUEST_CODE_GROUP) {
            if (data == null) {
                return;
            }
            mChoosenGroup = data.getStringExtra(ChooseGroupActivity.EXTRA_GROUP);
            UserSettings.setGroup(getActivity(), mChoosenGroup);
            mGroupPreference.setSummary(mChoosenGroup);
        }
    }

    @Override
    public void onStop() {
        super.onStop();
        UserSettings.setLinksPreference(getActivity(), mLinks.getValue());
        UserSettings.setSubjectsUpdateFrequency(getActivity(), mFregSubjects.getValue());
        UserSettings.setExamsUpdateFrequency(getActivity(), mFregExams.getValue());
    }
}

File: app/src/main/java/com/mai/nix/maiapp/navigation_fragments/settings/SettingsFragment.kt
package com.mai.nix.maiapp.navigation_fragments.settings

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.*
import android.widget.Toast
import androidx.appcompat.app.AppCompatDelegate
import androidx.browser.customtabs.CustomTabsIntent
import androidx.core.content.ContextCompat
import androidx.lifecycle.lifecycleScope
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import com.mai.nix.maiapp.MaiApp
import com.mai.nix.maiapp.R
import com.mai.nix.maiapp.choose_groups.NewChooseGroupActivity
import com.mai.nix.maiapp.helpers.UserSettings
import com.mai.nix.maiapp.helpers.UserSettings.setGroup
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.launch

/**
 * Created by Nix on 03.08.2017.
 */

@ExperimentalCoroutinesApi
class SettingsFragment : PreferenceFragmentCompat(), Preference.OnPreferenceClickListener, Preference.OnPreferenceChangeListener {
    private lateinit var groupPreference: Preference
    private lateinit var chosenGroup: String

    companion object {
        private const val REQUEST_CODE_GROUP = 0
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = super.onCreateView(inflater, container, savedInstanceState)
        view?.setBackgroundColor(ContextCompat.getColor(requireContext(), R.color.backgroundSplash))
        return view
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.app_prefs, rootKey)
        groupPreference = findPreference("pref_group")!!
        val clearSubjectsCache = findPreference<Preference>("clear_cache_subj")
        val clearExamsCache = findPreference<Preference>("clear_cache_ex")
        val frequencySubjects = findPreference<androidx.preference.ListPreference>("frequency_update_subjects")
        val frequencyExams = findPreference<androidx.preference.ListPreference>("frequency_update_exams")
        val theme = findPreference<androidx.preference.ListPreference>("pref_theme")
        val about = findPreference<Preference>("about")
        val mai = findPreference<Preference>("go_mai")

        groupPreference.summary = UserSettings.getGroup(requireContext())
        frequencySubjects?.value = UserSettings.getSubjectsUpdateFrequency(requireContext())
        frequencyExams?.value = UserSettings.getExamsUpdateFrequency(requireContext())

        groupPreference.onPreferenceClickListener = this
        clearSubjectsCache?.onPreferenceClickListener = this
        clearExamsCache?.onPreferenceClickListener = this
        about?.onPreferenceClickListener = this
        mai?.onPreferenceClickListener = this
        frequencySubjects?.onPreferenceChangeListener = this
        frequencyExams?.onPreferenceChangeListener = this
        theme?.onPreferenceChangeListener = this
    }

    private fun convertThemeValue(value: Any): Int {
        val i = value as String
        return i.toInt()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == REQUEST_CODE_GROUP && resultCode == Activity.RESULT_OK) {
            chosenGroup = data?.getStringExtra(NewChooseGroupActivity.EXTRA_GROUP)
                    ?: throw Exception("No group found")
            setGroup(requireContext(), chosenGroup)
            clearSubjectsCache(showMessage = false)
            clearExamsCache(showMessage = false)
            groupPreference.summary = chosenGroup
            requireActivity().setResult(Activity.RESULT_OK)
        }
    }

    override fun onPreferenceClick(preference: Preference): Boolean {
        when (preference.key) {
            "pref_group" -> {
                val i = NewChooseGroupActivity.newIntent(requireContext(), true)
                startActivityForResult(i, REQUEST_CODE_GROUP)
            }
            "clear_cache_subj" -> {
                clearSubjectsCache(showMessage = true)
            }
            "clear_cache_ex" -> {
                clearExamsCache(showMessage = true)
            }
            "about" -> {
                val intent = Intent(Intent.ACTION_SENDTO)
                intent.data = Uri.parse("mailto:novembernix@gmail.com?subject=Приложение МАИ")
                if (intent.resolveActivity(requireActivity().packageManager) != null) {
                    startActivity(intent)
                } else {
                    Toast.makeText(requireContext(), R.string.mail_error, Toast.LENGTH_SHORT).show()
                }
            }
            "go_mai" -> {
                val builder = CustomTabsIntent.Builder()
                builder.setShowTitle(true)
                builder.setToolbarColor(ContextCompat.getColor(requireContext(), R.color.colorText))
                val customTabsIntent = builder.build()
                customTabsIntent.launchUrl(requireContext(), Uri.parse("http://mai.ru/"))
            }
        }
        return true
    }

    private fun clearExamsCache(showMessage: Boolean) {
        lifecycleScope.launch {
            try {
                (requireContext().applicationContext as MaiApp).getDatabase().examDao.clearAll()
                if (showMessage) Toast.makeText(activity, R.string.clear_cache_exams_toast, Toast.LENGTH_SHORT).show()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    private fun clearSubjectsCache(showMessage: Boolean) {
        lifecycleScope.launch {
            try {
                (requireContext().applicationContext as MaiApp).getDatabase().scheduleDao.deleteAll()
                if (showMessage) Toast.makeText(activity, R.string.clear_cache_subj_toast, Toast.LENGTH_SHORT).show()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    override fun onPreferenceChange(preference: Preference, newValue: Any): Boolean {
        when (preference.key) {
            "frequency_update_subjects" -> {
                UserSettings.setSubjectsUpdateFrequency(requireContext(), newValue as String)
            }
            "frequency_update_exams" -> {
                UserSettings.setExamsUpdateFrequency(requireContext(), newValue as String)
            }
            "pref_theme" -> {
                val value = convertThemeValue(newValue)
                UserSettings.setTheme(requireContext(), value)
                AppCompatDelegate.setDefaultNightMode(value)
            }
        }
        return true
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        menu.clear()
        super.onCreateOptionsMenu(menu, inflater)
    }
}